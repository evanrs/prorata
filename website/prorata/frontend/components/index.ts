export * from './field'
export * from './currency-field'
export * from './investor-request-form'
