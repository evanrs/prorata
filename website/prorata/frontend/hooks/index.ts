export * from './use-mounted-effect'
export * from './use-gated-effect'
export * from './use-async-state'
export * from './use-layout-effect'
export * from './use-auto-focus'
