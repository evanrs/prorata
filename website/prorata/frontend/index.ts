export * from './tools'
export * from './hooks'
export * from './components'
export * from './prorata'
