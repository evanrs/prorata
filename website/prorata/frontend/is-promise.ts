export function isPromise<T>(value: unknown): value is Promise<T> {
  return value != null && typeof (value as Promise<T>)?.then === 'function'
}
