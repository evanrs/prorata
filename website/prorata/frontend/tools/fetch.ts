import { fetch, TypedResponse } from '@evanrs/fetch'

export interface ExtendedRequestInit<T> extends RequestInit {
  data?: T
}

export interface ExtendedResponse<T> extends TypedResponse<T> {
  data: T
}

async function __fetch<D, T>(
  request: RequestInfo,
  { data, headers, body, ...options }: ExtendedRequestInit<D> = {},
): Promise<ExtendedResponse<T>> {
  if (data != null) {
    headers = { ...headers, 'content-type': 'application/json' }
    body = JSON.stringify(data)
  }

  const response = await fetch<T>(
    `${process.env.API_BASE_URL ?? ''}${(request as Request)?.url ?? request}`,
    {
      ...options,
      headers,
      body,
    },
  )

  return Object.assign(response, {
    data: await response.json(),
  })
}

function isRequest(request: RequestInfo): request is Request {
  return String(request) === '[object Request]'
}

export { __fetch as fetch }
