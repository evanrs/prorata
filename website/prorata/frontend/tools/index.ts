export * from './fetch'
export * from './storage'
export * from './is-promise'
export * from './is-equal'
export * from './opacity-transition-for'
