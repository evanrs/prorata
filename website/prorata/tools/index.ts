export * from './fetch'
export * from './use-call-effect'
