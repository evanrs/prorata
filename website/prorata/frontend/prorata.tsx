import React, { useEffect, useState, useCallback, EffectCallback } from 'react'
import { Flex, Grid, GridProps, Heading } from '@chakra-ui/react'
import { useDebounce } from 'use-debounce'

import { Storage } from '.'
import { ajv, AllocationRequest, AllocationResponse } from '../common'

import { CurrencyField, InvestorRequestForm, InvestorUpdateHandler } from './components'

export type ProrataProps = {
  allocations?: AllocationResponse['allocations']
  allocationFor: (data?: AllocationRequest) => ReturnType<EffectCallback>
}

const isAllocationRequest = ajv.compile<AllocationRequest>(AllocationRequest)
const AllocationRequestStorage = Storage<AllocationRequest>('allocation-request')

export function Prorata({ allocations, allocationFor }: ProrataProps): JSX.Element {
  const [ready, setReady] = useState(false)
  const [allocation_amount, setAllocationAmount] = useState<number>()
  const [investor_amounts, setInvestorAmounts] = useState<AllocationRequest['investor_amounts']>([])

  const [autoFocused, focused] = useDebounce(
    !ready ? null : !allocation_amount ? 'allocation' : 'investor-request-form',
    3000,
  )
  useEffect(focused.flush, [ready])

  useEffect(() => {
    setReady(true)

    const request = AllocationRequestStorage.get('session')
    if (request) {
      setAllocationAmount(request?.allocation_amount)
      setInvestorAmounts(request?.investor_amounts ?? [])
    }
    // if we unmount let's clear our session
    return () => AllocationRequestStorage.remove('session')
  }, [])

  useEffect(() => {
    if (ready) {
      const request = { allocation_amount, investor_amounts }
      // store current session
      AllocationRequestStorage.set('session', request)
      // if its empty reset the allocations
      if (allocations?.length && investor_amounts.length === 0) {
        allocationFor()
      }
      // if its a valid request get the allocation
      else if (isAllocationRequest(request)) {
        allocationFor(request)
      }
    }
  }, [ready, allocation_amount, investor_amounts])

  const onInvestorUpdate: InvestorUpdateHandler = useCallback((name, value) => {
    setInvestorAmounts((investor_amounts) => {
      if (name === 'delete') {
        investor_amounts = investor_amounts.filter((v) => v !== value)
      } else if (value) {
        const index = name === 'new' ? investor_amounts.length : name
        investor_amounts = [
          ...investor_amounts.slice(0, index),
          value,
          ...investor_amounts.slice(index + 1),
        ]
      }

      return investor_amounts
    })
  }, [])

  return (
    <Flex direction="column">
      <Table my={0}>
        <Heading size="xs" my={1}>
          Total Available Allocation
        </Heading>
        <div />
        <div />
        <Heading size="xs" my={1} opacity={allocations?.length ? 1 : 0}>
          Total Allocated
        </Heading>
      </Table>
      <Table>
        <CurrencyField
          autoFocus={autoFocused === 'allocation'}
          placeholder="Allocation"
          name="allocation_amount"
          min={1}
          variant={autoFocused === 'allocation' || !allocation_amount ? 'outline' : 'filled'}
          value={allocation_amount ? allocation_amount : ''}
          set={(_, value) => {
            setAllocationAmount(
              typeof value === 'number'
                ? value
                : typeof value === 'string'
                ? Number(value)
                : undefined,
            )
          }}
        />
        <div />
        <div />

        <CurrencyField
          name="total"
          variant="flushed"
          readOnly
          value={allocations?.reduce((a, b) => a + b.allocation, 0)}
          opacity={allocations?.length ? 1 : 0}
        />

        <div />
      </Table>

      {/*  investor request form headings */}
      <Table my={0} mt=".5rem">
        <Heading size="xs" mt={4} mb={1}>
          Investor Breakdown
        </Heading>
        <Heading size="xs" mt={4} mb={1}>
          {/* Requested */}
        </Heading>
        <Heading size="xs" mt={4} mb={1}>
          {/* Average */}
        </Heading>
        <Heading
          size="xs"
          mt={4}
          mb={1}
          textAlign="left"
          opacity={investor_amounts?.length ? 1 : 0}
        >
          Investor Stake
        </Heading>
        <div />
      </Table>

      {/*  existing investor request forms */}
      {investor_amounts?.map((investor, i) => (
        <InvestorRequestForm
          key={`${investor.name}:${i}`}
          name={i}
          request={investor}
          allocation={allocations?.[i]}
          onUpdate={onInvestorUpdate}
        />
      ))}

      {/*  new investor request form */}
      <InvestorRequestForm
        // TODO drop the use of key here when handoff is complete
        key={`new:${investor_amounts.length}`}
        name="new"
        onUpdate={onInvestorUpdate}
        autoFocus={autoFocused === 'investor-request-form'}
      />
    </Flex>
  )
}

const Table: React.FC<GridProps> = (props) => (
  <Grid
    my={2}
    gap={[1, 1, 2, 2]}
    templateColumns="1fr 1fr 1fr minmax(4.5rem, .75fr) 3rem"
    alignItems="flex-end"
    {...props}
  />
)
