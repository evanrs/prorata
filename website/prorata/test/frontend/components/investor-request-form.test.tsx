import React from 'react'
import currency from 'currency.js'
import '@testing-library/jest-dom/extend-expect'

import { InvestorRequestForm, InvestorUpdateHandler } from '../../../frontend'
import { render, fireEvent, waitFor, screen } from '../../tools'
import { data } from '../../data'

describe('Investor Request Form', () => {
  it('matches snapshot', async () => {
    const request = data[0].request.investor_amounts[0]
    const onUpdate: InvestorUpdateHandler = (name, request) =>
      console.error(`onUpdate has not been mocked and received:`, name, request)

    const result = render(
      <InvestorRequestForm name={0} request={request} onUpdate={onUpdate} />,
      {},
    )

    expect(result.asFragment()).toMatchSnapshot()
  })

  it('populates fields', async () => {
    const request = data[0].request.investor_amounts[0]
    const onUpdate: InvestorUpdateHandler = (name, request) =>
      console.error(`onUpdate has not been mocked and received:`, name, request)

    render(<InvestorRequestForm name={0} request={request} onUpdate={onUpdate} />, {})

    const name = await inputFor('Name')
    expect(name.value).toEqual(request.name)

    const requested = await inputFor('Requested Amount')
    expect(currency(requested.value).value).toEqual(request.requested_amount)

    const average = await inputFor('Average Amount')
    expect(currency(average.value).value).toEqual(request.average_amount)
  })
})

function inputFor(placeholder: string) {
  return waitFor(() => screen.getByPlaceholderText(placeholder) as HTMLInputElement)
}
