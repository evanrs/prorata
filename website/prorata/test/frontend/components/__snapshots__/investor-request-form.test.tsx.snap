// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Investor Request Form matches snapshot 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    class="css-1l4fvs0"
  >
    <input
      autocomplete="false"
      hidden=""
      name="hidden"
    />
    <input
      autocomplete="off"
      class="chakra-input css-taf3kg"
      name="name"
      placeholder="Name"
      value="Investor A"
    />
    <div
      class="chakra-numberinput css-e69dqy"
      value="$100"
    >
      <input
        aria-valuemax="9007199254740991"
        aria-valuemin="0"
        aria-valuenow="100"
        aria-valuetext="$100"
        autocomplete="off"
        autocorrect="off"
        class="chakra-numberinput__field css-1o8sjxt"
        inputmode="numeric"
        name="requested_amount"
        pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
        placeholder="Requested Amount"
        role="spinbutton"
        type="text"
        value="$100"
        variant="filled"
      />
    </div>
    <div
      class="chakra-numberinput css-e69dqy"
      value="$100"
    >
      <input
        aria-valuemax="9007199254740991"
        aria-valuemin="0"
        aria-valuenow="100"
        aria-valuetext="$100"
        autocomplete="off"
        autocorrect="off"
        class="chakra-numberinput__field css-1o8sjxt"
        inputmode="numeric"
        name="average_amount"
        pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
        placeholder="Average Amount"
        role="spinbutton"
        type="text"
        value="$100"
        variant="filled"
      />
    </div>
    <div
      class="chakra-numberinput css-68rv0r"
      value=""
    >
      <input
        aria-readonly="true"
        aria-valuemax="9007199254740991"
        aria-valuemin="-9007199254740991"
        autocomplete="off"
        autocorrect="off"
        class="chakra-numberinput__field css-1o8sjxt"
        inputmode="numeric"
        name="allocation"
        pattern="^[+-]?[$]?[0-9]{1,3}(?:,?[0-9]{3})*(?:\\\\.[0-9]{2})?$"
        placeholder="Stake"
        readonly=""
        role="spinbutton"
        type="text"
        value=""
        variant="filled"
      />
    </div>
    <button
      class="chakra-button css-1fv4ia5"
      type="button"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
          />
        </g>
      </svg>
    </button>
    <input
      hidden=""
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;
